platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java

```diff
@@ -0,0 +1,322 @@
+   * will be updated on disk if necessary.
+   */
+  private void doHandleStore(ConfigurationContext context) throws IOException {
+    String pid = context.getServicePid();
+    File fileFromConfigAdmin = context.getConfigFile();
+
+    CachedConfigData cachedConfigData = pidDataMap.get(pid);
+    File fileFromCache = (cachedConfigData != null) ? cachedConfigData.getFelixFile() : null;
+
+    if (fileFromCache == null && fileFromConfigAdmin == null) {
+      // This config doesn't have an etc file, so we ignore this case (1)
+      LOGGER.debug("Was not tracked and will not track pid {}", pid);
+      return;
+    }
+
+    String appropriatePid =
+        (context.getFactoryPid() == null) ? context.getServicePid() : context.getFactoryPid();
+
+    if (fileFromCache == null) {
+      // An etc config file was just dropped and we're seeing it for the first time (2)
+      LOGGER.debug("Tracking pid {}", pid);
+      CachedConfigData createdConfigData = new CachedConfigData(context);
+      pidDataMap.put(pid, createdConfigData);
+      writeIfNecessary(
+          appropriatePid, fileFromConfigAdmin, context.getSanitizedProperties(), createdConfigData);
+      return;
+    }
+
+    if (!Objects.equals(fileFromConfigAdmin, fileFromCache)) {
+      if (fileFromConfigAdmin != null) {
+        // The felix file prop changed, which is not allowed (3)
+        String msg =
+            format(
+                "Felix filename has been illegally changed from [%s] to [%s]",
+                fileFromCache, fileFromConfigAdmin);
+        throw new IllegalStateException(msg);
+      }
+      if (fileFromCache.exists()) {
+        // The felix file prop was removed, which we can revert if the file exists (3)
+        LOGGER.info("Felix filename has been illegally removed, reverting to [{}]", fileFromCache);
+        context.setProperty(FELIX_FILENAME, fileFromCache.getAbsolutePath());
+        writeIfNecessary(
+            appropriatePid, fileFromCache, context.getSanitizedProperties(), cachedConfigData);
+        return;
+      }
+    }
+
+    // Routine property updates for tracked files - write to disk if necessary (4)
+    writeIfNecessary(
+        appropriatePid, fileFromConfigAdmin, context.getSanitizedProperties(), cachedConfigData);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Remove config files from etc when a delete occurs but only if felix is tracking the file.
+   *
+   * @param pid the pid pointing to the configuration to delete.
+   */
+  @Override
+  public void handleDelete(String pid) throws IOException {
+    try {
+      AccessController.doPrivileged(
+          (PrivilegedExceptionAction<Object>)
+              () -> {
+                doHandleDelete(pid);
+                return null;
+              });
+    } catch (PrivilegedActionException e) {
+      // Only caught if the action threw a CHECKED exception, thus it MUST be an IOException.
+      throw (IOException) e.getCause();
```

Comment:
❗️ As per a previous comment, we shouldn't blindly cast even if we know by design these are the only exceptions that are thrown. That side can evolve differently and we would then have a problem here.